core/arch/arm/tee/entry_std.c

```diff
@@ -327,11 +366,112 @@ static void entry_cancel(struct thread_smc_args *smc_args,
 	smc_args->a0 = OPTEE_SMC_RETURN_OK;
 }
 
+static void register_shm(struct thread_smc_args *smc_args,
+			 struct optee_msg_arg *arg, uint32_t num_params)
+{
+	paddr_t *pages = NULL;
+	int num_pages, pages_cnt;
+	struct optee_msg_param *params = OPTEE_MSG_GET_PARAMS(arg);
+
+	if (num_params == 0) {
+		arg->ret = TEE_ERROR_BAD_PARAMETERS;
+		return;
+	}
+
+	num_pages = (params[0].u.tmem.size +
+		     (params[0].u.tmem.buf_ptr & SMALL_PAGE_MASK) - 1)
+		/ SMALL_PAGE_SIZE + 1;
+
+	pages = malloc(num_pages * sizeof(paddr_t));
+
+	if (!pages) {
+		arg->ret = TEE_ERROR_OUT_OF_MEMORY;
+		return;
+	}
+
+	pages_cnt = extract_pages_from_params(params, pages, num_params);
+	if (pages_cnt < 0) {
+		arg->ret = TEE_ERROR_BAD_PARAMETERS;
+		goto out;
+	}
+
+	if (num_pages != pages_cnt) {
+		arg->ret = TEE_ERROR_BAD_PARAMETERS;
+		goto out;
+	}
+
+	if (mobj_reg_shm_alloc(pages, pages_cnt,
+			       params[0].u.tmem.shm_ref) == NULL) {
+		arg->ret = TEE_ERROR_GENERIC;
+		goto out;
+	}
+
+	arg->ret = TEE_SUCCESS;
+out:
+	free(pages);
+	smc_args->a0 = OPTEE_SMC_RETURN_OK;
+}
+
+static void unregister_shm(struct thread_smc_args *smc_args,
+			   struct optee_msg_arg *arg, uint32_t num_params)
+{
+	struct optee_msg_param *params = OPTEE_MSG_GET_PARAMS(arg);
+
+	if (num_params == 1) {
+		struct mobj *mobj;
+		uint64_t cookie = params[0].u.rmem.shm_ref;
+
+		mobj = mobj_reg_shm_find_by_cookie(cookie);
+		if (mobj) {
+			mobj_free(mobj);
+			arg->ret = TEE_SUCCESS;
+		} else {
+			EMSG("Can't find mapping with given cooike\n");
+			arg->ret = TEE_ERROR_BAD_PARAMETERS;
+		}
+	} else {
+		arg->ret = TEE_ERROR_BAD_PARAMETERS;
+		arg->ret_origin = TEE_ORIGIN_TEE;
+	}
+
+	smc_args->a0 = OPTEE_SMC_RETURN_OK;
+}
+
+static struct shmem_mapping *map_cmd_buffer(paddr_t parg)
+{
+	struct shmem_mapping *shm_mapping = NULL;
+	struct optee_msg_arg *arg = NULL;	/* fix gcc warning */
+	uint32_t num_params;
+	size_t args_size;
+
+	if (shmem_map_page(parg, &shm_mapping))
+		return NULL;
+
+	arg = (struct optee_msg_arg *)shmem_pa2va(parg);
+	if (arg == NULL)
+		goto err;
+
+	num_params = arg->num_params;
```

Comment:
`num_params` will be read once again in `tee_entry_std()` below after this function has returned. What if normal world succeeds in using this window to increase `num_params`?